src/codesign.py

```diff
@@ -188,13 +206,61 @@ def inverse_pass(self):
         self.symbolic_sim.calculate_edp(self.hw)
         self.symbolic_sim.save_edp_to_file()
 
-        self.inverse_edp = self.symbolic_sim.edp.subs(self.tech_params)
+        # print(f"tech params: {self.tech_params}")
+        # print(f"type of tech_param keys: {type(list(self.tech_params.keys())[0])}")
+
+        # for symbol in self.symbolic_sim.edp.free_symbols:
+        #     for key in self.tech_params.keys():
+        #         # print(f"key: {key}")
+        #         # print(f"symbol: {symbol}")
+        #         if key.name == symbol.name:
+        #             print(f"key: {key} == symbol: {symbol}")
+        #             if key != symbol:
+        #                 print(f"{key}: {key.assumptions0}")
+        #                 print(f"{symbol}: {symbol.assumptions0}")
+        #         # else:
+                #     print(f'key: {key} != symbol: {symbol}')
+            # assert symbol in self.tech_params.keys()
+
+        # tech_params2 = {}
+        # for key in self.tech_params.keys():
+        #     tech_params2[key.name] = self.tech_params[key]
+
+        # print(f"tech params 2: {tech_params2}")
+        inverse_cycles = self.symbolic_sim.cycles.xreplace(self.tech_params).evalf()
+        print(f"free symbols in cycles after substition: {inverse_cycles.free_symbols}")
+
+        print(f"cycles: {inverse_cycles}")
+
+        self.inverse_edp = self.symbolic_sim.edp.xreplace(self.tech_params).evalf()
```

Comment:
confused by the order here. Why is EDP calculated before active and passive energy? I guess the edp expression already has all the expression subs done on it....

Let's actually verify that the result is the same when we do edp.evalf()
vs `delay.evalf() * (active.evalf() + passive.evalf())`

Reply:
Unsure. I don't think I added this.